### ISSUE TITLE
index query that utilizes joins result in wrong $model parameter for add policies

### ISSUE THREAD

**Issue by sedlatschek** [laravel/nova-issues#3373]
- Laravel Version: 8.38.0
- Nova Version: 3.23.2
- PHP Version: 7.4.0

### Description:

Using a custom resource index query that utilizes joins results in a wrong $model parameter for add policies.
Is it a bad practice to use joins in index queries?

### Detailed steps to reproduce the issue on a fresh Nova installation:

1. Add model & resource `User`
2. Add model & resource `Organisation`
3. Add model & resource `Branch`
4. Add model & resource `Department`
5. Add model & resource `Something`
5. Setup relations:
- `Department` n:1 `Branch`
- `Branch` n:1 `Organisation`
- `User` n:1 `Organisation`
- `Something` n:1 `Department`

5. Add `indexQuery` to `Deparment` resource
```php
    // App\Nova\Department
    public static function indexQuery(Request $request, $query)
    {
        return $query
                ->leftJoin('branches', 'departments.branch_id', '=', 'branches.id')
                ->where('branches.organisation_id', $request->user()->organisation_id)
                ->select('departments.*');
    }
```

6. Add `DepartmentPolicy`
```php
    // App\Policies\DepartmentPolicy
    public function addSomething(User $user, Department $department)
    {
        // $department does not have the actual values of the department,
        // it is filled with values from the branches table
    }
```

The reason for this seems to be the `passes` function in `Rules\Relatable.php`:
```php
    public function passes($attribute, $value)
    {
        $model = $this->query->tap(function ($query) {
            $query->getQuery()->orders = [];
        })->select('*')->whereKey($value)->first();

        // the select('*') replaces the custom select('departments.*'), thats the reaseon
        // the policy parameter $department for addSomething has values from branches.
        // deleting `->select('*')` (for testing purposes) solves the problem.

        if (! $model) {
            return false;
        }

        if ($this->relationshipIsFull($model, $attribute, $value)) {
            return false;
        }

        if ($resource = Nova::resourceForModel($model)) {
            return $this->authorize($resource, $model);
        }

        return true;
    }
```

**Comment by crynobone** [comment]
Can you create full reproducing app based on above example? I'm not sure I can't follow the example just based on what's available above.

**Comment by sedlatschek** [comment]
Hey, here is the example: https://github.com/sedlatschek/laravel-nova-issue-3373

**Comment by crynobone** [comment]
Why does `App\Nova\Department` only have `BelongsTo` field to `Branch` but then have `addSomething` policy. Shouldn't it be `addBranch` policy?

**Comment by sedlatschek** [comment]
`Something` belongs to `Department`, thats why `Department` has an `addSomething` policy

**Comment by crynobone** [comment]
Also, I personally would think `whereHas` is better for this sort of query. 

```php
public static function indexQuery(Request $request, $query)
{
    return $query->whereHas('branches', function ($query) use ($request) {
        return $query->where((new \App\Models\Branch()->getQualifiedKeyName()), $request->user()->organisation_id);
    });
}
``

**Comment by crynobone** [comment]
> Something belongs to Department

But it's not declared as a field under `Department` resource. That's one issue.

**Comment by sedlatschek** [comment]
> > Something belongs to Department
> 
> But it's not declared as a field under `Department` resource. That's one issue.

Am I missing something? Isn't this the relation you mean?
https://github.com/sedlatschek/laravel-nova-issue-3373/blob/master/app/Nova/Something.php#L47

**Comment by crynobone** [comment]
`App\Policies\Department` only handle policy for `Department` resource or relationship constraint for `App\Nova\Department`, and not `App\Nova\Something`.

**Comment by sedlatschek** [comment]
> `App\Policies\Department` only handle policy for `Department` resource or relationship constraint for `App\Nova\Department`, and not `App\Nova\Something`.

But why is it fired when I update a `Something` that belongs to a `Department`.

**Comment by crynobone** [comment]
> The docs clearly show that a policy can affect other models:

Yes it does, but it can only handle `addSomething` for `HasMany::make('Something')` on `Department` resource.

**Comment by sedlatschek** [comment]
Sorry, I really don't want to be an asshole, I just don't get the point.
`addSomething` already fires and it still does when I add the hasMany relation: https://github.com/sedlatschek/laravel-nova-issue-3373/commit/55a36eeb7107fe288d19648271efe95b5c308ffb

You're right, using `whereHas` can serve as a workaround, but using joins in the `indexQuery` function still affects $model parameter in the `add` policies. Thats the bug I am reporting.

**Comment by crynobone** [comment]
I don't think we will be able to support this. You should be using `whereHas()` or following (which is identical):

```php
public static function indexQuery(Request $request, $query)
{
    return $query->whereExists(function($query) use ($request) {
        return $query->select('*')
              ->from('branches')
              ->whereRaw('departments.branch_id=branches.id')
              ->where('branches.organisation_id', '=', $request->user()->organisation_id);
    });
}
```

-----

> deleting `->select('*')` (for testing purposes) solves the problem.

We cannot do this as if `indexQuery()` set a subset `->select()` we might be missing some attributes which will be needed for `Rules\Relatable.php`.

**Comment by sedlatschek** [comment]
Ok, thanks for your help.

Deleting `->select('*')` was not meant as suggestion, it was just to show that this very piece of code causes the problem.

**Comment by sedlatschek** [comment]
This was fixed with [3.25.0](https://nova.laravel.com/releases/3.25.0).

The new `passes` function in `Relatable.php` sets a prefix:
```php
        $model = $this->query->tap(function ($query) {
            tap($query->getQuery(), function ($builder) {
                $builder->orders = [];

                $builder->select(
                    ! empty($builder->joins) ? $builder->from.'.*' : '*'
                );
            });
        })->whereKey($value)->first();
```

Thanks a lot :)
